### ISSUE TITLE
Yeoman generates files with broken permissions

### ISSUE THREAD

**Issue by superghostmaster** [yeoman/generator-angular#978]
Your default instructions result in a generated app that has folder dependencies which are completely mismatched to the actual folder locations.

Generating yo angular newapp results in Bootstrap not even being loaded because it's in a directory which doesn't reside where the index.html is calling it from.

Running grunt generates an equally, but differently broken "dist" version of the app which also has missing dependencies.

Your instructions suck.

**Comment by superghostmaster** [comment]
Loading the default index file, pulled from Chrome console:

--
 XMLHttpRequest cannot load file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/views/main.html. Cross origin requests are only supported for protocol schemes: http, data, chrome-extension, https, chrome-extension-resource.
vendor.24764a48.js:5 Error: Failed to execute 'send' on 'XMLHttpRequest': Failed to load 'file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/views/main.html'.
    at Error (native)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:151
    at v (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:4:30289)
    at j (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:4:28812)
    at i (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:13850)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:14022
    at l.$eval (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:21149)
    at l.$digest (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:19658)
    at l.$apply (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:21416)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:3:26563
vendor.24764a48.js:5 Error: [$compile:tpload] Failed to load template: views/main.html
http://errors.angularjs.org/1.3.9/$compile/tpload?p0=views%2Fmain.html
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:3:20441
    at g (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:27265)
    at i (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:13850)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:14022
    at l.$eval (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:21149)
    at l.$digest (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:19658)
    at l.$apply (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:5:21416)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:3:26563
    at Object.e [as invoke] (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:4:5935)
    at h (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/dist/scripts/vendor.24764a48.js:3:26484)
file://www.google-analytics.com/analytics.js Failed to load resource: net::ERR_FILE_NOT_FOUND

**Comment by superghostmaster** [comment]
Loading the yo-generated index file:
--
Failed to load resource: net::ERR_FILE_NOT_FOUND
angular.js:9814 XMLHttpRequest cannot load file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/views/main.html. Cross origin requests are only supported for protocol schemes: http, data, chrome-extension, https, chrome-extension-resource.
angular.js:11594 Error: Failed to execute 'send' on 'XMLHttpRequest': Failed to load 'file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/views/main.html'.
    at Error (native)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:9814:11
    at sendReq (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:9615:9)
    at serverRequest (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:9331:16)
    at processQueue (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:13171:27)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:13187:27
    at Scope.$eval (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:14384:28)
    at Scope.$digest (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:14200:31)
    at Scope.$apply (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:14489:24)
    at bootstrapApply (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:1449:15)
angular.js:11594 Error: [$compile:tpload] Failed to load template: views/main.html
http://errors.angularjs.org/1.3.9/$compile/tpload?p0=views%2Fmain.html
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:63:12
    at handleError (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:16040:17)
    at processQueue (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:13171:27)
    at file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:13187:27
    at Scope.$eval (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:14384:28)
    at Scope.$digest (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:14200:31)
    at Scope.$apply (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:14489:24)
    at bootstrapApply (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:1449:15)
    at Object.invoke (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:4182:17)
    at doBootstrap (file:///Users/CodeMachine/Documents/Boilerplates/test_yeoman/newapp/app/bower_components/angular/angular.js:1447:14)

**Comment by eddiemonge** [comment]
What instructions are you following? What folder output are you seeing? What are you using as the webserver because you can't just open index.html in a browser. You need to run the grunt server or another properly configured (for the setup this generator creates) server.

Also, your title was extremely hostile and not at all help-inducing. In fact I had to take a couple of deep breaths to calm myself before responding.

**Comment by passy** [comment]
@superghostmaster You are the reason why people get burned out in open source. Don't be such a prick.

**Comment by sindresorhus** [comment]
Happy to hear about your pain points, but keep in mind that your tone is proportionate to how much we'll care about your feedback.

**Comment by hemanth** [comment]
![](http://gifs.joelglovier.com/wtf/enough-internet.gif)

**Comment by superghostmaster** [comment]
This is after running the index file *after* grunt.

There are clear issues with folder dependency, and as the original post says:  running grunt creates another copy of the entire app in a subfolder.  Running THAT index file fails the same way, albeit folder paths now changed due to being in a subfolder.

Running yet ANOTHER grunt in the subfolder creates yet another folder containing the entire app again.

**Comment by superghostmaster** [comment]
After explicit, careful review of each of the errors I've seen, and a quick number of Google searches resulting in the error, "Error: EACCES, permission denied '/var/root/.config' " constantly leading back to posts specifically regarding yeoman, an angular in particular, I can say it appears the issue is two-fold:

1)  You're trying to write to /var/root, which under OS X, even under sudo su the user doesn't have access to write to.

Please don't suggest chowning my root folder as this is a terrible, insecure and horrible solution, though I've seen it suggested multiple times under the github account associated w/this generator.

2)  You're also trying to write to *cached* NPM folders as part of an app being generated.  This will almost always result in permissions issues.  

The simple fix is to clear the npm cache (npm clean cache) but this presents yet another problem:  the app and generator seem to take ownership of the npm cache that they create.  This presents the issue that any other npm requests that depend on these folders now lose their permissions, requiring another npm cache clean.  No biggy, but a pain in the ass none the less and one that's unnecessary.

A simple fix would be to create your npm dependency folder one time, and cache under /.npm/generator-angular/{module name} but instead, for things like, hmm... angular?? you try to write directly to /.npm/angular.

This *appears* (I haven't tested in detail to confirm because I've now wiped yeoman from my system as troubleshooting yeoman is still taking longer than simply generating with a few extra commands) to result in the app not finding it's library dependencies it expects to pull from npm because they were never created and/or it doesn't own the perms to see them.

Chowning folders and files *after* the install of yeoman and/or generation of an app shouldn't be necessary if the generator is properly configuring the folders and files in the first place.  With this in mind, I decided to check out the perms on some of the YAML files that it was failing on (yet another error) and found a rather amateur mistake:  You give read permissions to r/w to System and for Everyone you leave it as No Access, which obviously results in the file being inaccessible to the app.

Why this happens I didn't spend time investigating, mostly because I've tried this under about every possible permissions scenario already and the same thing happens.  All I know is your generator is creating files without the proper permissions, wether or not npm, yeoman, generator-angular or any other part of the process is install with root, sudo su or the user.

**Comment by eddiemonge** [comment]
There shouldn't be any permission errors unless you are running `yo` with sudo (or as root initially).

Nothing in the generator would try to change ownership of files in the npm cache directory.

What do you mean by `running the index.html` file. How exactly are you `running` it?

What version of `yo` and `generator-angular` are/were you using?

**Comment by eddiemonge** [comment]
closing due to inactivity

**Comment by superghostmaster** [comment]
Just wanted to report back - @eddiemonge 's solution, re: npm was installed using sudo was ultimately the cause of my pain.

I solved this by completely removing all node-related folders, etc.  and installing w/out sudo.

For those who might run into the same issue, I found a pretty straight-forward way of making sure you do this correctly, via Mr. John Papa (who has more than once come to my rescue):
http://www.johnpapa.net/how-to-use-npm-global-without-sudo-on-osx/

TL;DR - do this:
```
brew install node --without-npm
mkdir "${HOME}/.npm-packages"
echo NPM_PACKAGES="${HOME}/.npm-packages" >> ${HOME}/.bashrc
echo prefix=${HOME}/.npm-packages >> ${HOME}/.npmrc
curl -L https://www.npmjs.org/install.sh | sh
echo NODE_PATH=\"\$NPM_PACKAGES/lib/node_modules:\$NODE_PATH\" >> ${HOME}/.bashrc
echo PATH=\"\$NPM_PACKAGES/bin:\$PATH\" >> ${HOME}/.bashrc
echo source "~/.bashrc" >> ${HOME}/.bash_profile
source ~/.bashrc
```

This basically removes any garbage left over from node but doesn't fully remove all of your node folders, which may still have associations in your $PATH.  Doing a search and trash was the only method I could find to really wipe them all.  It's not pretty but if you make a stupid move like myself and install node @ root/sudo then it's worth the extra 10 min or so.

The commands above will effectively associate node so that sudo is unnecessary so that perms errors you get when other packages install (which assume you don't have node installed requiring sudo) should now be a non-issue.
